src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+        /// Unlike the <see cref="JsonUtf8Reader"/>, which is a ref struct, the state can survive
+        /// across async/await boundaries and hence this type is required to provide support for reading
+        /// in more data asynchronously before continuing with a new instance of the <see cref="JsonUtf8Reader"/>.
+        /// </summary>
+        public JsonReaderState CurrentState => new JsonReaderState
+        {
+            _stackFreeContainer = _stackFreeContainer,
+            _lineNumber = _lineNumber,
+            _lineBytePosition = _lineBytePosition,
+            _bytesConsumed = BytesConsumed,
+            _currentDepth = _currentDepth,
+            _maxDepth = _maxDepth,
+            _inObject = _inObject,
+            _isNotPrimitive = _isNotPrimitive,
+            _tokenType = _tokenType,
+            _readerOptions = _readerOptions,
+            _stack = _stack,
+            _sequencePosition = Position,
+        };
+
+        /// <summary>
+        /// Constructs a new <see cref="JsonUtf8Reader"/> instance.
+        /// </summary>
+        /// <param name="jsonData">The ReadOnlySpan&lt;byte&gt; containing the UTF-8 encoded JSON text to process.</param>
+        /// <param name="isFinalBlock">True when the input span contains the entire data to process.
+        /// Set to false only if it is known that the input span contains partial data with more data to follow.</param>
+        /// <param name="state">If this is the first call to the ctor, pass in a default state. Otherwise,
+        /// capture the state from the previous instance of the <see cref="JsonUtf8Reader"/> and pass that back.</param>
+        /// <remarks>
+        /// Since this type is a ref struct, it is a stack-only type and all the limitations of ref structs apply to it.
+        /// This is the reason why the ctor accepts a <see cref="JsonReaderState"/>.
+        /// </remarks>
+        public JsonUtf8Reader(ReadOnlySpan<byte> jsonData, bool isFinalBlock, JsonReaderState state)
+        {
+            _buffer = jsonData;
+
+            _isFinalBlock = isFinalBlock;
+
+            // Note: We do not retain _bytesConsumed or _sequencePosition as they reset with the new input data
+            _stackFreeContainer = state._stackFreeContainer;
+            _lineNumber = state._lineNumber;
+            _lineBytePosition = state._lineBytePosition;
+            _currentDepth = state._currentDepth;
+            _maxDepth = state._maxDepth == 0 ? JsonReaderState.StackFreeMaxDepth : state._maxDepth;
+            _inObject = state._inObject;
+            _isNotPrimitive = state._isNotPrimitive;
+            _tokenType = state._tokenType;
+            _readerOptions = state._readerOptions;
+            _stack = state._stack;
+
+            _consumed = 0;
+            _totalConsumed = 0;
+            _isLastSegment = _isFinalBlock;
+            _isSingleSegment = true;
+            _currentPosition = default;
+            //_nextPosition = default;
```

Comment:
Uncomment or delete, but please don't have commented out code (except when it's part of a descriptive comment, like "I'm not doing `code` because `reason`", or "Need to `what this does` to support `an enlightening caller code example`"